mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/PassChainExecutor.scala

```diff
@@ -14,37 +14,37 @@ import at.forsyte.apalache.tla.lir.MissingTransformationError
  * @author Igor Konnov
  */
 
-class PassChainExecutor @Inject() (val options: WriteablePassOptions,
-    @Named("InitialPass") val initialPass: Pass with TlaModuleMixin)
-    extends LazyLogging {
+class PassChainExecutor(val options: WriteablePassOptions, passes: Seq[Pass]) extends LazyLogging {
 
-  def run(): Option[Pass with TlaModuleMixin] = {
-    def exec(seqNo: Int, passToRun: Pass with TlaModuleMixin): Option[Pass with TlaModuleMixin] = {
+  def run(): Option[TlaModule] = {
+    def exec(seqNo: Int, passToRun: Pass, module: TlaModule): Option[TlaModule] = {
       logger.info("PASS #%d: %s".format(seqNo, passToRun.name))
-      val result = passToRun.execute()
-      val outcome = if (result) "[OK]" else "[FAIL]"
+      val result = passToRun.execute(module)
+      val outcome = if (result.isDefined) "[OK]" else "[FAIL]"
       logger.debug("PASS #%d: %s %s".format(seqNo, passToRun.name, outcome))
-      if (!result) {
-        None // return the negative result
-      } else {
-        val nextPass = passToRun.nextPass
-        if (nextPass.hasModule) {
-          val module = nextPass.tlaModule.get
-
-          // Raise error if the pass dependencies aren't satisfied
-          if (!nextPass.dependencies.subsetOf(module.properties)) {
-            val missing = nextPass.dependencies -- module.properties
-            throw new MissingTransformationError(
-                s"${nextPass.name} cannot run for a module without the properties: ${missing.mkString(", ")}", module)
-          }
-
-          exec(1 + seqNo, nextPass) // call the next pass in line
-        } else {
-          Some(passToRun) // finished
-        }
-      }
+      result
     }
 
-    exec(0, initialPass)
+    var module = new TlaModule("empty", Seq()) with TlaModuleProperties
+    passes.zipWithIndex.foreach { case (pass, index) =>
+      // Raise error if the pass dependencies aren't satisfied
+      if (!pass.dependencies.subsetOf(module.properties)) {
+        val missing = pass.dependencies -- module.properties
+        throw new MissingTransformationError(
+            s"${pass.name} cannot run for a module without the properties: ${missing.mkString(", ")}", module)
+      }
+
+      val transformedModule = exec(index, pass, module)
+
+      transformedModule match {
+        case None => return None
+        case Some(m) =>
+          val newModule = new TlaModule(m.name, m.declarations) with TlaModuleProperties
+          newModule.properties = module.properties ++ pass.transformations
+
+          module = newModule
+      }
+    }
+    Some(module)
```

Comment:
```suggestion
    passes.zipWithIndex.foldLeft(Option[TlaModule with TlaModuleProperties](new TlaModule("empty", Seq()) with TlaModuleProperties)){ case (moduleOpt, (pass, index)) =>
      // Raise error if the pass dependencies aren't satisfied
      if (!pass.dependencies.subsetOf(module.properties)) {
        val missing = pass.dependencies -- module.properties
        throw new MissingTransformationError(
            s"${pass.name} cannot run for a module without the properties: ${missing.mkString(", ")}", module)
      }
      for {
        module <- moduleOpt
        transformedModule <- exec(index, pass, module)
      } yield {
        val newModule = new TlaModule(transformedModule .name, transformedModule .declarations) with TlaModuleProperties
        newModule.properties = module.properties ++ pass.transformations
        newModule
      }
    }
```

Reply:
You might need to tweak this a bit before committing, because I wrote it in the github editor, but the basic idea is that 
```
for {
  x <- optionX // optionX: Option[A]
  y <- f(x) // f: A -> Option[A]
} yield y // None if optionX or f(x) are None, Some(y) otherwise
```
is a neat way to handle fold + options.